pkg/symbolicexpr/symbolicexpr_test.go

```diff
@@ -0,0 +1,152 @@
+	conjSrc, conjDst, conjEmpty := Conjunction{}, Conjunction{}, Conjunction{}
+	for i := 1; i <= 3; i++ {
+		testTag := initTestTag(fmt.Sprintf("t%v", i))
+		atomic := &atomicTerm{label: testTag, toVal: fmt.Sprintf("str%v", i)}
+		conjSrc = *conjSrc.add(atomic)
+		negateAtomic := atomic.negate().(atomicTerm)
+		conjDst = *conjDst.add(&negateAtomic)
+	}
+	conjSymbolicPath := SymbolicPath{conjSrc, conjDst}
+	fmt.Printf("\nconjSymbolicPath:\n%v\n", conjSymbolicPath.string())
+	require.Equal(t, "(t1 = str1 and t2 = str2 and t3 = str3) to (t1 != str1 and t2 != str2 and t3 != str3)",
+		conjSymbolicPath.string(), "conjSymbolicPath not as expected")
+	println("conjEmpty", conjEmpty.string())
+	require.Equal(t, emptySet, conjEmpty.string(), "empty conjunction not as expected")
+}
+
+func TestComputeAllowGivenDenySingleTermEach(t *testing.T) {
+	conjSrc1, conjDst1, conjSrc2, conjDst2 := Conjunction{}, Conjunction{}, Conjunction{}, Conjunction{}
+	testSrc1 := initTestTag("s1")
+	atomic1 := &atomicTerm{label: testSrc1, toVal: "str1"}
+	conjSrc1 = *conjSrc1.add(atomic1)
+	testDst1 := initTestTag("d1")
+	atomicDst1 := &atomicTerm{label: testDst1, toVal: "str1"}
+	conjDst1 = *conjDst1.add(atomicDst1)
+	testSrc2 := initTestTag("s2")
+	atomic2 := &atomicTerm{label: testSrc2, toVal: "str2"}
+	conjSrc2 = *conjSrc2.add(atomic2)
+	testDst2 := initTestTag("d2")
+	atomicDst2 := &atomicTerm{label: testDst2, toVal: "str2"}
+	conjDst2 = *conjDst2.add(atomicDst2)
+	allowPath := SymbolicPath{conjSrc1, conjDst1}
+	denyPath := SymbolicPath{conjSrc2, conjDst2}
+	fmt.Printf("allowPath is %v\ndenyPath is %v\n", allowPath.string(), denyPath.string())
+	allowGivenDeny := *computeAllowGivenDeny(allowPath, denyPath)
+	fmt.Printf("computeAllowGivenDeny(allowPath, denyPath) is\n%v\n", allowGivenDeny.string())
+	require.Equal(t, "(s1 = str1 and s2 != str2) to (d1 = str1)\n(s1 = str1) to (d1 = str1 and d2 != str2)",
+		allowGivenDeny.string(), "allowGivenDeny single term computation not as expected")
+}
+
+func TestComputeAllowGivenDenyThreeTermsEach(t *testing.T) {
+	conjAllow, conjDeny := Conjunction{}, Conjunction{}
+	for i := 1; i <= 3; i++ {
+		testAllow := initTestTag(fmt.Sprintf("s%v", i))
+		atomicAllow := &atomicTerm{label: testAllow, toVal: fmt.Sprintf("str%v", i)}
+		conjAllow = *conjAllow.add(atomicAllow)
+		testDeny := initTestTag(fmt.Sprintf("s%v`", i))
+		atomicDeny := &atomicTerm{label: testDeny, toVal: fmt.Sprintf("str%v`", i)}
+		conjDeny = *conjDeny.add(atomicDeny)
+	}
+	allowPath := SymbolicPath{conjAllow, conjAllow}
+	denyPath := SymbolicPath{conjDeny, conjDeny}
+	fmt.Printf("symbolicAllow is %s\nsymbolicDeny is %s\n", allowPath.string(), denyPath.string())
+	allowGivenDeny := *computeAllowGivenDeny(allowPath, denyPath)
+	fmt.Printf("computeAllowGivenDeny(allowPath, denyPath) is\n%v\n", allowGivenDeny.string())
+	require.Equal(t,
+		"(s1 = str1 and s2 = str2 and s3 = str3 and s1` != str1`) to (s1 = str1 and s2 = str2 and s3 = str3)\n"+
```

Comment:
what does "to" in the expr means?

Reply:
can avoid using hard coded strings in tests? it will be very difficult to maintain this way.

Reply:
how can one capture input and output per test?

Reply:
let's first merge and handle these later.

Reply:
#67